### PR TITLE
feat: add osgi dependency manager

### LINKED ISSUE #18861: Add support updating dependencies in OSGi Feature Model models
### What would you like Renovate to be able to do?

# New package manager questionnaire

**Did you read our documentation on adding a package manager?**

- [x] I've read the [adding a package manager](adding-a-package-manager.md) documentation.

### Name of package manager

OSGi Feature Model

### What language does this support?

Java 

### How popular is this package manager?

The OSGi Feature model ( [OSGi specification](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.feature.html) , [Reference guide](https://sling.apache.org/documentation/development/feature-model.html) ) supports building OSGi applications. It has a base implementation and a rich set of tooling in the Apache Sling project ( https://github.com/apache/?q=sling+feature&type=all&language=&sort= ). It is used by applications built on top of Apache Sling and also by "pure" OSGi applications.

### Does this language have other (competing?) package managers?

 The OSGi Feature Model supports resolution of artifacts from Maven repositories. If I understand the question correctly, Maven and Gradle are competing package managers. 

- [x] Yes - Maven, Gradle
- [ ] No

---

### What type of package files and names does it use?

OSGi Feature Model files are JSON files, with certain sections designating Maven dependencies. These files are usually placed under `src/main/features`  for projects using existing tooling.

### What [fileMatch](../usage/configuration-options.md/#filematch) pattern(s) should be used?

```^src\/main\/features\/.*\.json$```

### Is it likely that many users would need to extend this pattern for custom file names?

- [ ] Yes
- [x] No

### Is the fileMatch pattern likely to get many "false hits" for files that have nothing to do with package management?

No

---

### Can package files have "local" links to each other that need to be resolved?

No. While there is a feature to "extend" a feature file, the artifacts found in a single feature file can be processed independently.

### Is there a reason why package files need to be parsed together (in serial) instead of independently?

N/A

### What format/syntax is the package file in?

- [x] JSON
- [ ] TOML
- [ ] YAML
- [ ] Custom (explain below)

### How do you suggest parsing the file?

- [x] Off the shelf parser
- [ ] Using regex
- [ ] Custom-parsed line by line
- [ ] Other

### Does the package file structure distinguish between different "types" of dependencies? e.g. production dependencies, development dependencies, etc?

- [ ] Yes, production and development dependencies
- [x] No, all dependencies are treated the same

#### Plain listing of artifacts

```
"bundles": [
    "org.apache.sling:org.apache.sling.api:2.10.1",
    "org.apache.sling:org.apache.sling.engine:2.0.0"
]
```

#### Object listing of artifactrs

```
"bundles": [
    { "id": "org.apache.sling:org.apache.sling.api:2.10.1" },
    { "id": "org.apache.sling:org.apache.sling.engine:2.0.0" }
]
```

#### Alternative listings of artifacts

There are two extensions.

The first one is in the reference implementation, where the _bundles_ property can be replaced by _section-name:ARTIFACTS|(true|false)_.

The second one is just in the spec, and follows the form

```
    "extensions": {
        "org.acme.ddlfiles": {
            "kind": "mandatory",
            "type": "artifacts",
            "artifacts": [
              { "id": "org.acme:appddl:1.2.1" }, 
              {
                "id": "org.acme:appddl-custom:1.0.3",
                "org.acme.target": "custom-db"
              }
           ]
        }
    }
```

### Describe which types of dependencies above are supported and which will be implemented in future

If I understand the question correctly - there is no implementation at the moment. My contribution will contain the first two forms: _bundles_, and _section-name:ARTIFACTS|(true|false)_. 

I plan to add to support for the "extensions" variant as soon as that lands in the reference implementation.

---

### What versioning scheme does the package file(s) use?

Artifacts are defined in the compact Maven format: `groupId ':' artifactId ( ':' type ( ':' classifier )? )? ':' version`

### Does this versioning scheme support range constraints, e.g. `^1.0.0` or `1.x`?

- [ ] Supports range constraints (e.g `^1.0.0` or `1.x`)
- [x] No

---

### Is a new datasource required? Provide details

- [ ] Yes, provide details.
- [x] No.

### Will users need the capability to specify a custom host/registry to look up? Can it be found within the package files, or within other files inside the repository, or would it require Renovate configuration?

There is no need to specify a custom host/registry. The mechanism should be the same one as used by Maven/Gradle.

### Do the package files have any "constraints" on the parent language (e.g. supports only v3.x of Python) or platform (Linux, Windows, etc.) that should be used in the lookup procedure?

There are no constraints on the parent language or platform.

### Will users need the ability to configure language or other constraints using Renovate config?

I am not sure I understand the question, but there should be no need to have additional configuration, on top of what is already possible for Maven/Gradle.

---

### Are lock files or checksum files used? Are they mandatory?

No lockfiles are used. Checksums are used if Maven/Gradle are configured to do so, so only for downloads.

### If so, what tool and exact commands should be used if updating one or more package versions in a dependency file?

N/A

### If applicable, describe how the tool maintains a cache and if it can be controlled via CLI or environment variables? Do you recommend the cache be kept or disabled/ignored?

Maven maintains a local repository cache, by default in `$HOME/.m2/repository`. It is recommended that this cache is enabled.

### If applicable, what command should be used to generate a lock file from scratch if you already have a package file? This will be used for "lock file maintenance"

N/A

### Is there anything else to know about this package manager?

Not that I can think of.

### If you have any ideas on how this should be implemented, please tell us here.

It should be enough to parse the JSON files for sections that are known to contain artifacts and extract those.

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by rombert**
## Changes

Adds a new dependency manager - `osgi`.

## Context

Closes #18861

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/osgifeature/extract.ts
Comment (secustor) [inline lib/modules/manager/osgifeature/extract.ts:56]:
```suggestion
    logger.warn(`Failed parsing ${fileName}: ${(e as Error).message}`);
```
For better readability.

File: lib/modules/manager/osgifeature/extract.ts
Comment (secustor) [inline lib/modules/manager/osgifeature/extract.ts:61]:
```suggestion
  return deps.length ? { deps } : null;
```

File: lib/modules/manager/osgifeature/extract.ts
Comment (secustor) [inline lib/modules/manager/osgifeature/extract.ts:67]:
```suggestion
  if ('bundles' === sectionName || sectionName.includes(':ARTIFACTS|')) {
```

File: lib/modules/manager/osgifeature/index.ts
Comment (secustor) [inline lib/modules/manager/osgifeature/index.ts:16]: This is only necessary if it differs from the module name. 
```suggestion
```

File: lib/modules/manager/osgifeature/index.ts
Comment (secustor) [inline lib/modules/manager/osgifeature/index.ts:11]:
```suggestion
```
This will be set by the `maven` datasource.

**Review by secustor — CHANGES_REQUESTED** [review]
Thanks for your contribution! 

Some thoughts and improvements.

File: lib/modules/manager/osgifeature/extract.ts
Comment (rombert) [reply lib/modules/manager/osgifeature/extract.ts:24]: e4daa1ff73ab20221891ec02f4b967a6c6c0106e

File: lib/modules/manager/api.ts
Comment (rombert) [reply lib/modules/manager/api.ts:144]: I am hesitant to claim the "osgi" dependency type with my contribution. In OSGi there is no single standard for dependecy formats, and the one I am implementing is one of the recent proposals.

For instance, we have the [Karaf Features](https://karaf.apache.org/manual/latest/provisioning) or the now-deprecated [Sling Provisioning Model](https://sling.apache.org/documentation/development/slingstart.html) which serve a similar purpose.

That being said, if you're fine with this, I can change the name to `osgi`, just wanted to bring all the information to the discussion.

File: lib/modules/manager/osgifeature/extract.ts
Comment (viceice) [reply lib/modules/manager/osgifeature/extract.ts:24]: please read this: https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#resolve-review-comments-instead-of-commenting

File: lib/modules/manager/osgifeature/extract.ts
Comment (rombert) [reply lib/modules/manager/osgifeature/extract.ts:20]: @secustor - by a typed approach do you mean using an interface/class to represent the complete feature model definition? I id not do that for two reasons:

1. The model is extensible, with dynamic property names, which makes it hard to do proper typing. As you can see in the PR, we check for certain section names
2. The model is richer than just describing dependencies, it can also hold configurations, text section ( e.g. DDL for an SQL database ) and I think representing all of that is not needed

File: lib/modules/manager/osgifeature/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/osgifeature/readme.md:10]:
```suggestion
The `osgifeature` manager extracts dependencies from feature model definition files, typically located under `src/main/features`.
It uses the `maven` datasource to find dependency updates.

Artifact list extensions are not supported.
For the definition of artifact list extensions, read [section 159.7.3 of the OSGi R8 spec](INSERT LINK TO SECTION HERE).

References:

- [OSGi compendium release 8, Feature Service Specification](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.feature.html)
- [The Sling implementation of the feature model](https://sling.apache.org/documentation/development/feature-model.html)
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some docs changes, remember to insert the link to the section after applying my suggestion. 😉

File: lib/modules/manager/osgifeature/extract.ts
Comment (HonkingGoose) [inline lib/modules/manager/osgifeature/extract.ts:1]: In the README for this feature I see that the OSGi compendium is on release `8`. What happens when OSGi releases version `9`, with breaking changes?

Maybe the code should check which version of OSGi it uses, and throw an error if it's a newer version which we don't support?

Comment (rombert) [reply lib/modules/manager/osgifeature/extract.ts:1]: The implementations referenced in the OSGI specs rarely make breaking changes.

There is a provision for versioning at https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.feature.html#d0e156846, but no actual implementation uses it. The versioning part works by declaring an attribute of the form

```
"feature-resource-version": "1.0",
```

I can read it, and fail if it exists and does not "semver-match" `^1`.  Does that sound ok?

File: lib/modules/manager/osgifeature/extract.ts
Comment (secustor) [reply lib/modules/manager/osgifeature/extract.ts:20]: You would have to add only the attributes you actually need for Renovate. 

e.g. 
```ts
export interface MyModel {
  bundles?: Bundle[];
  'execution-environment:JSON|false'?: {
    'framework?': Bundle
  }
  [x: string]: any;
}

export type Bundle = string | BundleObject;

export interface BundleObject {
  id: string;
}
``` 

But I'm fine to approve it within the current scope if the whole json is typed, but the processing should be typed anyway.

File: lib/modules/manager/osgifeature/extract.ts
Comment (secustor) [inline lib/modules/manager/osgifeature/extract.ts:38]:
```suggestion
      for (const entry of extractArtifactList(section, value)) {
        const rawGav = typeof entry === 'string' ? entry : entry.id

        // both '/' and ':' are valid separators, but the Maven datasource
        // expects the separator to be ':'
        const gav = rawGav.replaceAll('/', ':');
```

File: lib/modules/manager/osgifeature/extract.ts
Comment (secustor) [inline lib/modules/manager/osgifeature/extract.ts:70]:
```suggestion
function extractArtifactList(sectionName: string, sectionValue: any): Bundle[] {
  // Compendiun R8 159.4: bundles entry
  // The 'ARTIFACTS' key is supported by the Sling/OSGi feature model implementation
  if ('bundles' === sectionName || sectionName.includes(':ARTIFACTS|')) {
    return sectionValue as Bundle[];
  }
```

File: lib/modules/manager/osgifeature/types.ts
Comment (secustor) [inline lib/modules/manager/osgifeature/types.ts:3]:
```suggestion
export type Bundle = string | BundleObject;

export interface BundleObject {
  id: string;
}
```

File: lib/modules/manager/osgifeature/__fixtures__/double-slash-not-comment.json
Comment (secustor) [inline lib/modules/manager/osgifeature/__fixtures__/double-slash-not-comment.json:4]: Please also add some malfunction tests like missing `id` attributes inside of bundles.

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/osgifeature/extract.ts
Comment (HonkingGoose) [reply lib/modules/manager/osgifeature/extract.ts:1]: It sounds good to me! 😉

File: lib/modules/manager/api.ts
Comment (secustor) [reply lib/modules/manager/api.ts:144]: 👍

File: lib/modules/manager/osgifeature/extract.ts
Comment (secustor) [reply lib/modules/manager/osgifeature/extract.ts:20]: There is no way for fully dynamic ones, but templating is possible. 

```ts
type keys = 'content-packages' | 'other-packages' 
export type Artifacts = `${keys}:ARTIFACTS|${boolean}`

// the same as 
export type Templated = `content-packages:ARTIFACTS|true` |
                        `content-packages:ARTIFACTS|false` |
                        `other-packages:ARTIFACTS|true` |
                        `other-packages:ARTIFACTS|false` |
```

You will have to filter entries, based on your conditions.

File: lib/modules/manager/osgi/extract.ts
Comment (secustor) [inline lib/modules/manager/osgi/extract.ts:104]: Please follow https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#logging

File: lib/modules/manager/osgi/extract.ts
Comment (secustor) [inline lib/modules/manager/osgi/extract.ts:113]: Same as above. https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#logging

**Review by secustor — CHANGES_REQUESTED** [review]

**Review by secustor — DISMISSED** [review]
LGTM. 👍     
Requesting now a maintainer review.

File: lib/modules/manager/osgi/__fixtures__/bundles-as-strings.json
Comment (viceice) [inline lib/modules/manager/osgi/__fixtures__/bundles-as-strings.json:1]: please inline those small fixtures to the test

File: lib/modules/manager/osgi/types.ts
Comment (viceice) [inline lib/modules/manager/osgi/types.ts:7]:
```suggestion
  [x: string]: unknown;
```

File: lib/modules/manager/osgi/extract.ts
Comment (viceice) [inline lib/modules/manager/osgi/extract.ts:121]:
```suggestion
function extractArtifactList(sectionName: string, sectionValue: unknown): Bundle[] {
```

File: lib/modules/manager/osgi/extract.ts
Comment (viceice) [inline lib/modules/manager/osgi/extract.ts:35]:
```suggestion
  const allBundles = featureModel.bundles ?? [];
```
i don't think we need the explicit type

File: lib/modules/manager/osgi/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/osgi/extract.spec.ts:8]: move fixtures inside test if only used once

File: lib/modules/manager/osgi/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/osgi/extract.spec.ts:7]: why these aren't inlined?

File: lib/modules/manager/osgi/extract.ts
Comment (viceice) [inline lib/modules/manager/osgi/extract.ts:30]: featureModel can be null here when the json is `null` which is valid json5, so validate with `is.plainObject` too.

File: lib/modules/manager/osgi/extract.ts
Comment (viceice) [inline lib/modules/manager/osgi/extract.ts:123]: probably validata value with `is.array`

File: lib/modules/manager/osgi/extract.spec.ts
Comment (rombert) [reply lib/modules/manager/osgi/extract.spec.ts]: I misread the previous comments - I thought I should've moved only "small" fixtures.

Comment (viceice) [reply lib/modules/manager/osgi/extract.spec.ts]: they where all small 🙃

File: lib/modules/manager/osgi/extract.ts
Comment (rombert) [reply lib/modules/manager/osgi/extract.ts:34]: I had to revert this change in 5a0108222 as it was making the linter unhappy. I am not sure exactly what is going on, but it seems that a check like

```javascript
if (!is.plainObject(featureModel)) {
  return null
}
```

or even 
```javascript
if (!is.plainObject<FeatureModel>(featureModel)) {
  return null
}
```

affects the effective type of the `featureModel` variable. I'm open to suggestions on how to fix this in a more elegant manner, but for now I simply check for null and undefined.

**Comment by rombert** [comment]
@rarkins  - I fixed node 14 compatiblity with 7aa8752

**Comment by rombert** [comment]
@rarkins - `extract.ts` now has 100% coverage, can you please re-trigger a check?

File: lib/modules/manager/osgi/extract.ts
Comment (rarkins) [inline lib/modules/manager/osgi/extract.ts:25]:
```suggestion
  } catch (err) {
    logger.warn({ fileName, err }, 'Failed to parse osgi file');
```

For warnings and errors, we want the `message` to be static so that it can be aggregated or ignored in people's log systems.

If this is likely to happen often and be ignorable then it could just be a debug message instead of warn.

File: lib/modules/manager/osgi/extract.ts
Comment (rarkins) [inline lib/modules/manager/osgi/extract.ts:54]:
```suggestion
    logger.trace({ fileName, section }, 'Parsing section');
```

File: lib/modules/manager/osgi/extract.ts
Comment (rarkins) [inline lib/modules/manager/osgi/extract.ts:17]:
```suggestion
  logger.trace({ fileName }, 'osgi.extractPackageFile');
```

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/osgi/extract.ts
Comment (secustor) [inline lib/modules/manager/osgi/extract.ts:34]:
```suggestion
  if (
    // for empty an empty result
    is.nullOrUndefined(featureModel) ||
    // Compendium R8 159.9: resource versioning
    !isSupportedFeatureResourceVersion(featureModel, fileName)
  ) {
```
I can confirm for me `is.plainObject(value)` is also not working me either. It returns `true` for every objet generated by the JSON5 library.

**Review by secustor — CHANGES_REQUESTED** [review]

**Review by secustor — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Comment by rombert** [comment]
@rarkins - is there anything left for me to do here? I see two approvals, not sure if I should request other reviews.

**Comment by rarkins** [comment]
It needs to be updated to be up to date with main branch

**Comment by rombert** [comment]
> It needs to be updated to be up to date with main branch

Wasn't aware, thanks. Updated.

**Pull request merged**
